File: packages/shared/src/protocol.ts
// ---------------------------------------------------------------------------
// @azul/shared — protocol.ts
// Zod schemas and inferred TypeScript types for the WebSocket wire protocol.
// All client→server and server→client messages are discriminated unions on
// the `type` field. parseClientMessage validates incoming raw data at runtime.
// ---------------------------------------------------------------------------

import { z } from 'zod';

// Inline the colour list to avoid a circular-import with index.ts
// (index.ts re-exports everything from protocol.ts, so protocol.ts must
// not import from index.ts at module evaluation time).
const COLORS = ['blue', 'yellow', 'red', 'black', 'white'] as const;

// ---------------------------------------------------------------------------
// Bot types — defined locally to avoid a shared→engine dependency cycle.
// ---------------------------------------------------------------------------

/**
 * AI difficulty level. Mirrors the same type in @azul/engine/bot/types.ts (structurally identical).
 * Kept separate to avoid a shared↔engine dependency cycle.
 * The compile-time mutual-assignability guard lives in packages/engine/src/bot/bot-types.test.ts.
 */
export type BotLevel = 'easy' | 'medium' | 'hard';
const BotLevelSchema = z.enum(['easy', 'medium', 'hard']);

// ---------------------------------------------------------------------------
// Reusable sub-schemas (shared between client and server messages)
// ---------------------------------------------------------------------------

const ColorSchema = z.enum(COLORS);

/** Move source: either a specific factory display or the centre pool. */
const MoveSourceSchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('factory'), index: z.number().int().nonnegative() }),
  z.object({ type: z.literal('center') }),
]);

/** Move target: a pattern line row (0–4) or the floor line. */
const MoveTargetSchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('patternLine'), row: z.number().int().min(0).max(4) }),
  z.object({ type: z.literal('floor') }),
]);

/** A single game move intent (mirrors the Move domain type). */
const MoveSchema = z.object({
  source: MoveSourceSchema,
  color: ColorSchema,
  target: MoveTargetSchema,
});

// ---------------------------------------------------------------------------
// Room interface (used in lobby payloads)
// ---------------------------------------------------------------------------

/**
 * A lobby room record, shared between the Room repository and the wire
 * protocol. `createdAt` is an ISO-8601 timestamp string.
 */
export interface Room {
  id: string;
  name: string;
  hostId: string;
  maxPlayers: number;
  players: { id: string; name: string; bot?: { level: BotLevel } }[];
  status: 'waiting' | 'playing' | 'finished';
  isPrivate: boolean;
  createdAt: string;
}

const RoomPlayerSchema = z.object({
  id: z.string(),
  name: z.string(),
  bot: z.object({ level: BotLevelSchema }).optional(),
});

const RoomSchema = z.object({
  id: z.string(),
  name: z.string(),
  hostId: z.string(),
  maxPlayers: z.number().int().min(2).max(4),
  players: z.array(RoomPlayerSchema),
  status: z.enum(['waiting', 'playing', 'finished']),
  isPrivate: z.boolean(),
  createdAt: z.string(),
});

// ---------------------------------------------------------------------------
// Client → Server message schemas
// ---------------------------------------------------------------------------

const HelloSchema = z.object({
  type: z.literal('hello'),
  /** Auth/reconnect token stored in localStorage. */
  token: z.string(),
  /** Present on reconnect to a specific room. */
  roomId: z.string().optional(),
});

const LobbySubscribeSchema = z.object({
  type: z.literal('lobby:subscribe'),
});

const RoomCreateSchema = z.object({
  type: z.literal('room:create'),
  name: z.string().min(1),
  /** 2–4 players per Azul rules. */
  maxPlayers: z.number().int().min(2).max(4),
  isPrivate: z.boolean(),
});

const RoomJoinSchema = z.object({
  type: z.literal('room:join'),
  roomId: z.string(),
});

const RoomLeaveSchema = z.object({
  type: z.literal('room:leave'),
  roomId: z.string(),
});

const RoomStartSchema = z.object({
  type: z.literal('room:start'),
  roomId: z.string(),
});

const GameMoveSchema = z.object({
  type: z.literal('game:move'),
  move: MoveSchema,
  /**
   * Client echoes the current turnSeq. Server rejects the move if this
   * doesn't match, preventing double-tap / stale-move races.
   */
  expectedTurnSeq: z.number().int(),
});

const PingSchema = z.object({
  type: z.literal('ping'),
});

const RoomAddBotSchema = z.object({
  type: z.literal('room:addBot'),
  roomId: z.string(),
  level: BotLevelSchema,
});

/**
 * Discriminated union of every valid client→server message.
 * The `type` field is the discriminant.
 */
export const ClientMessageSchema = z.discriminatedUnion('type', [
  HelloSchema,
  LobbySubscribeSchema,
  RoomCreateSchema,
  RoomJoinSchema,
  RoomLeaveSchema,
  RoomStartSchema,
  GameMoveSchema,
  PingSchema,
  RoomAddBotSchema,
]);

export type ClientMessage = z.infer<typeof ClientMessageSchema>;

// ---------------------------------------------------------------------------
// Server → Client message schemas
// ---------------------------------------------------------------------------

// PlayerView is a complex nested type from index.ts. For the wire-protocol
// schema we use z.unknown() as a pass-through because the engine already
// constructs and validates this shape; the schema is a structural placeholder
// that lets other workers type the payload correctly via the exported
// PlayerView interface from index.ts.
const PlayerViewSchema = z.unknown();

const HelloOkSchema = z.object({
  type: z.literal('hello:ok'),
  playerId: z.string(),
});

const LobbyStateSchema = z.object({
  type: z.literal('lobby:state'),
  rooms: z.array(RoomSchema),
});

const RoomStateSchema = z.object({
  type: z.literal('room:state'),
  room: RoomSchema,
});

const GameStateSchema = z.object({
  type: z.literal('game:state'),
  /** Redacted PlayerView — built by engine.toPlayerView(). */
  view: PlayerViewSchema,
});

const GameTurnSchema = z.object({
  type: z.literal('game:turn'),
  currentPlayerId: z.string(),
  /** Unix-ms deadline timestamp (server clock). */
  deadline: z.number(),
});

const GameAppliedSchema = z.object({
  type: z.literal('game:applied'),
  move: MoveSchema,
  /** PlayerId that made the move. */
  by: z.string(),
  turnSeq: z.number().int(),
});

const GameOverSchema = z.object({
  type: z.literal('game:over'),
  scores: z.array(z.object({ playerId: z.string(), score: z.number() })),
  winnerIds: z.array(z.string()),
});

const GameAbortedSchema = z.object({
  type: z.literal('game:aborted'),
  reason: z.string(),
});

const PlayerConnectionSchema = z.object({
  type: z.literal('player:connection'),
  playerId: z.string(),
  connected: z.boolean(),
});

const SessionInvalidSchema = z.object({
  type: z.literal('session:invalid'),
});

const PongSchema = z.object({
  type: z.literal('pong'),
});

const ErrorSchema = z.object({
  type: z.literal('error'),
  code: z.string(),
  message: z.string(),
});

/**
 * Discriminated union of every valid server→client message.
 * The `type` field is the discriminant.
 */
export const ServerMessageSchema = z.discriminatedUnion('type', [
  HelloOkSchema,
  LobbyStateSchema,
  RoomStateSchema,
  GameStateSchema,
  GameTurnSchema,
  GameAppliedSchema,
  GameOverSchema,
  GameAbortedSchema,
  PlayerConnectionSchema,
  SessionInvalidSchema,
  PongSchema,
  ErrorSchema,
]);

export type ServerMessage = z.infer<typeof ServerMessageSchema>;

// ---------------------------------------------------------------------------
// parseClientMessage — public entry point used by the WS gateway
// ---------------------------------------------------------------------------

/**
 * Parse and validate a raw (unknown) WebSocket message as a ClientMessage.
 * Throws a ZodError with descriptive path/message info on any validation
 * failure (invalid type, missing fields, out-of-range values, etc.).
 *
 * @param raw  Anything — typically `JSON.parse(ws.data)` output.
 * @returns    A fully-typed ClientMessage discriminated union value.
 */
export function parseClientMessage(raw: unknown): ClientMessage {
  return ClientMessageSchema.parse(raw);
}

File: azul-server/crates/shared/src/protocol.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::{Move, MoveTarget, PlayerView};

/// AI difficulty level. In the TS codebase this is duplicated in both
/// shared/protocol.ts and engine/bot/types.ts to avoid a shared<->engine
/// dependency cycle. In Rust, `azul-engine` depends on `azul-shared`, so the
/// engine bot reuses THIS type — there is no duplicate and the mutual
/// assignability guard test is obsolete.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, TS)]
#[serde(rename_all = "lowercase")]
#[ts(export)]
pub enum BotLevel {
    Easy,
    Medium,
    Hard,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "lowercase")]
#[ts(export)]
pub enum RoomStatus {
    Waiting,
    Playing,
    Finished,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct RoomPlayerBot {
    pub level: BotLevel,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct RoomPlayer {
    pub id: String,
    pub name: String,
    /// Present for AI players; absent for humans (`bot?` in TS -> omit when None).
    #[serde(skip_serializing_if = "Option::is_none", default)]
    // The wire omits `bot` for humans (skip_serializing_if); mirror that as an
    // optional TS field since ts-rs ignores the serde attr above.
    #[ts(optional)]
    pub bot: Option<RoomPlayerBot>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct Room {
    pub id: String,
    pub name: String,
    pub host_id: String,
    pub max_players: u32,
    pub players: Vec<RoomPlayer>,
    pub status: RoomStatus,
    pub is_private: bool,
    /// ISO-8601 timestamp string.
    pub created_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct ScoreEntry {
    #[serde(rename = "playerId")]
    pub player_id: String,
    pub score: i32,
}

/// Every valid client->server message. Internally tagged on `type`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(tag = "type")]
#[ts(export)]
pub enum ClientMessage {
    #[serde(rename = "hello")]
    Hello {
        token: String,
        #[serde(rename = "roomId", skip_serializing_if = "Option::is_none", default)]
        // ts-rs can't parse the combined serde attr above (it ignores it), so
        // restate the rename + optionality for codegen explicitly.
        #[ts(optional, rename = "roomId")]
        room_id: Option<String>,
    },
    #[serde(rename = "lobby:subscribe")]
    LobbySubscribe,
    #[serde(rename = "room:create")]
    RoomCreate {
        name: String,
        #[serde(rename = "maxPlayers")]
        max_players: u32,
        #[serde(rename = "isPrivate")]
        is_private: bool,
    },
    #[serde(rename = "room:join")]
    RoomJoin {
        #[serde(rename = "roomId")]
        room_id: String,
    },
    #[serde(rename = "room:leave")]
    RoomLeave {
        #[serde(rename = "roomId")]
        room_id: String,
    },
    #[serde(rename = "room:start")]
    RoomStart {
        #[serde(rename = "roomId")]
        room_id: String,
    },
    #[serde(rename = "game:move")]
    GameMove {
        #[serde(rename = "move")]
        mv: Move,
        #[serde(rename = "expectedTurnSeq")]
        #[ts(type = "number")]
        expected_turn_seq: i64,
    },
    #[serde(rename = "ping")]
    Ping,
    #[serde(rename = "room:addBot")]
    RoomAddBot {
        #[serde(rename = "roomId")]
        room_id: String,
        level: BotLevel,
    },
}

impl ClientMessage {
    /// Range/length constraints that serde alone does not enforce (these were
    /// zod refinements in the TS schema). Returns Err with a message on failure.
    pub fn validate(&self) -> Result<(), String> {
        match self {
            ClientMessage::RoomCreate {
                name, max_players, ..
            } => {
                if name.is_empty() {
                    return Err("name must be non-empty".into());
                }
                if !(2..=4).contains(max_players) {
                    return Err("maxPlayers must be between 2 and 4".into());
                }
            }
            ClientMessage::GameMove { mv, .. } => {
                if let MoveTarget::PatternLine { row } = mv.target
                    && row > 4
                {
                    return Err("patternLine row must be between 0 and 4".into());
                }
            }
            _ => {}
        }
        Ok(())
    }
}

/// Parse and validate a raw WebSocket text frame as a `ClientMessage`.
/// Mirrors `parseClientMessage` (JSON.parse + zod). Returns Err on any
/// malformed frame (bad `type`, missing field, out-of-range value).
pub fn parse_client_message(raw: &str) -> Result<ClientMessage, String> {
    let msg: ClientMessage = serde_json::from_str(raw).map_err(|e| e.to_string())?;
    msg.validate()?;
    Ok(msg)
}

/// Every valid server->client message. Internally tagged on `type`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[serde(tag = "type")]
#[ts(export)]
pub enum ServerMessage {
    #[serde(rename = "hello:ok")]
    HelloOk {
        #[serde(rename = "playerId")]
        player_id: String,
    },
    #[serde(rename = "lobby:state")]
    LobbyState { rooms: Vec<Room> },
    #[serde(rename = "room:state")]
    RoomState { room: Room },
    #[serde(rename = "game:state")]
    GameState { view: PlayerView },
    #[serde(rename = "game:turn")]
    GameTurn {
        #[serde(rename = "currentPlayerId")]
        current_player_id: String,
        /// Unix-ms deadline timestamp (server clock).
        #[ts(type = "number")]
        deadline: i64,
    },
    #[serde(rename = "game:applied")]
    GameApplied {
        #[serde(rename = "move")]
        mv: Move,
        by: String,
        #[serde(rename = "turnSeq")]
        #[ts(type = "number")]
        turn_seq: u64,
    },
    #[serde(rename = "game:over")]
    GameOver {
        scores: Vec<ScoreEntry>,
        #[serde(rename = "winnerIds")]
        winner_ids: Vec<String>,
    },
    #[serde(rename = "game:aborted")]
    GameAborted { reason: String },
    #[serde(rename = "player:connection")]
    PlayerConnection {
        #[serde(rename = "playerId")]
        player_id: String,
        connected: bool,
    },
    #[serde(rename = "session:invalid")]
    SessionInvalid,
    #[serde(rename = "pong")]
    Pong,
    #[serde(rename = "error")]
    Error { code: String, message: String },
}

impl ServerMessage {
    /// Serialize to a JSON text frame for the socket.
    pub fn to_json(&self) -> String {
        serde_json::to_string(self).expect("ServerMessage serialization cannot fail")
    }
}
